mount/losetup_linux.go

```diff
@@ -0,0 +1,180 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package mount
+
+import (
+	"fmt"
+	"math/rand"
+	"os"
+	"strings"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	loopControlPath = "/dev/loop-control"
+	loopDevFormat   = "/dev/loop%d"
+
+	// According to util-linux/include/loopdev.h
+	ioctlSetFd       = 0x4C00
+	ioctlClrFd       = 0x4C01
+	ioctlSetStatus64 = 0x4C04
+	ioctlGetFree     = 0x4C82
+
+	loFlagsReadonly = 1
+	//loFlagsUseAops   = 2
+	loFlagsAutoclear = 4
+	//loFlagsPartScan  = 8
+	loFlagsDirectIO = 16
+
+	ebusyString = "device or resource busy"
+)
+
+// struct loop_info64 in util-linux/include/loopdev.h
+type loopInfo struct {
+	/*
+		device         uint64
+		inode          uint64
+		rdevice        uint64
+		offset         uint64
+		sizelimit      uint64
+		number         uint32
+		encryptType    uint32
+		encryptKeySize uint32
+	*/
+	_        [13]uint32
+	flags    uint32
+	fileName [64]byte
+	/*
+		cryptName  [64]byte
+		encryptKey [32]byte
+		init       [2]uint64
+	*/
+	_ [112]byte
+}
+
+func ioctl(fd, req, args uintptr) (uintptr, uintptr, error) {
+	r1, r2, errno := syscall.Syscall(syscall.SYS_IOCTL, fd, req, args)
+	if errno != 0 {
+		return 0, 0, errno
+	}
+
+	return r1, r2, nil
+}
+
+func getFreeLoopDev() (uint32, error) {
```

Comment:
You might want to consider using `losetup` [package](https://github.com/containerd/containerd/tree/master/snapshots/devmapper/losetup), which was merged to `master` with devmapper snapshotter recently.

Reply:
@mxpv IMO, there are two main disadvantages of the `losetup` package as it is calling `losetup` binary directly:

1. There is no way to set `LO_FLAGS_AUTOCLEAR` flag with `losetup` binary, which makes the whole mount cleanup procedure complicated. For example, after calling `losetup` to setup a loopback device, we always have to call `losetup -d` to clean it up either after umount, or if anything goes wrong during mount. OTOH, in `losetup_linux.go`, I just need to set the `LO_FLAGS_AUTOCLEAR` flag when creating the loopback device, and do not worry about loopback device cleaning up at all.

2. `losetup` package requires `exec.Command` to fork a child process whereas `losetup_linux.go` does not fork and is thus a bit more efficient.

Reply:
After we merge this in, is there a way we can consolidate these implementations?

Reply:
@dmcgowan Yes, we can re-implement the same APIs exposed in the devmapper `losetup` package w/o explicitly calling `losetup` binary in `losetup_linux.go`, and change devmapper to use the generic `mount` package for the same APIs. The attach part is already here. I used to implement the detach part but deleted it as I don't really need it. I can work on the consolidation once this is merged.